### PR TITLE
Simplify loop condition

### PR DESCRIPTION
**Pull request by EikanWang**
Fixes https://github.com/pytorch/pytorch/pull/54121

### CODE REVIEW

File: torch/csrc/jit/tensorexpr/ir_simplifier.cpp [hunk 1]
```diff
@@ -2910,6 +2911,203 @@ ExprPtr SimplifierUnderContext::mutate(DivPtr v) {
...
+    int diff = immediateAs<int>(ret);
+    TORCH_INTERNAL_ASSERT(diff != 0);
+    if (diff > 0) {
+      // lhs = [5, 10] and rhs = [0, 4]
+      cmp_res = CompareSelectOperation::kGT;
+    } else {
+      // lhs = [5, 10] and rhs = [11, 15]
+      cmp_res = CompareSelectOperation::kLT;
+    }
+  }
```
Comment (jgong5) [inline torch/csrc/jit/tensorexpr/ir_simplifier.cpp:3062]: For the range comparison logic above, seems the `boundOverlap` doesn't help simply the logic a lot. Why not simply compare the bound1.end and bound2.start directly? You may create a help function like `maybeLessOrEqual(bound1, bound2)` on two bounds that check `bound1.end` and `bound2.start` and return either Less, LessEqual or Equal, or unknown if none of the three meets. Then you can invoke `maybeLessOrEqual` on `(lhs_bound, rhs_bound)` and `(rhs_bound, lhs_bound)` respectively. For the latter, you just need to negate on the result.

File: torch/csrc/jit/tensorexpr/ir_simplifier.cpp [hunk 2]
```diff
@@ -2910,6 +2911,203 @@ ExprPtr SimplifierUnderContext::mutate(DivPtr v) {
...
+          ? simp_ret1
+          : ((cmp_res == CompareSelectOperation::kGT ||
+              cmp_res == CompareSelectOperation::kLT)
+                 ? simp_ret2
+                 : simplified_cmp_select_expr);
+      break;
+    default:
+      TORCH_INTERNAL_ASSERT(false);
+      break;
+  }
```
Comment (jgong5) [inline torch/csrc/jit/tensorexpr/ir_simplifier.cpp:3105]: You may consider to simplify the logic by adding two operations on `CompareSelectOperation`:
1. `imply(CompareSelectOperation op1, CompareSelectOperation op2)` which implements the logic "op1 -> op2".
2. `not(CompareSelectOperation op)` which implements the logic "!op".

Then, you only need to check the result of `imply(cmp_res, v->compare_select_op())` and `imply(cmp_res, not(v->compare_select_op()))` to decide whether ret1 or ret2 should be returned.

File: torch/csrc/jit/tensorexpr/ir_simplifier.cpp [hunk 3]
```diff
@@ -2910,6 +2911,203 @@ ExprPtr SimplifierUnderContext::mutate(DivPtr v) {
...
+        return false;
+      }
+      auto got = var_bound_info.find(simple_expr_ptr);
+      if (got == var_bound_info.end()) {
+        return false;
+      } else {
+        bound_info.start = got->second.first;
+        // TODO: Need to add the range bounary information to Bound
+        bound_info.end = IRSimplifier::simplify(
+            alloc<Sub>(got->second.second, immLike(got->second.second, 1)));
```
[the remark this answers is not in the document]
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/ir_simplifier.cpp]: Yes

File: torch/csrc/jit/tensorexpr/ir_simplifier.cpp [hunk 4]
```diff
@@ -2910,6 +2911,203 @@ ExprPtr SimplifierUnderContext::mutate(DivPtr v) {
...
+  }
+
+  if (simp_condition->isConstant()) {
+    auto ret = immediateAs<int>(simp_condition);
+    if (ret) {
+      return simp_true_val;
+    } else {
+      return simp_false_val;
+    }
+  } else {
```
[the remark this answers is not in the document]
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/ir_simplifier.cpp]: Fixed.

File: torch/csrc/jit/tensorexpr/ir_simplifier.cpp [hunk 5]
```diff
@@ -2910,6 +2911,203 @@ ExprPtr SimplifierUnderContext::mutate(DivPtr v) {
...
+  ExprPtr false_val = v->false_value();
+
+  auto simp_condition = IRSimplifier::simplify(condition->accept_mutator(this));
+  auto simp_true_val = IRSimplifier::simplify(true_val->accept_mutator(this));
+  auto simp_false_val = IRSimplifier::simplify(false_val->accept_mutator(this));
+
+  ExprPtr simplified_ifelse_expr = nullptr;
+  if ((simp_condition == condition) && (simp_true_val == true_val) &&
+      (simp_false_val == false_val)) {
+    simplified_ifelse_expr = v;
```
[the remark this answers is not in the document]
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/ir_simplifier.cpp]: Fixed.

File: torch/csrc/jit/tensorexpr/ir_simplifier.cpp (hunk 1 above)
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/ir_simplifier.cpp:3070]: Yes, the `boundOverlap` does not help too much but we still can leverage part of its logic. And we should move the logic to `bounds_overlap.h` or `bounds_inference.h`. Currently, I abstract the logic as a lambda function.

File: torch/csrc/jit/tensorexpr/ir_simplifier.cpp (hunk 2 above)
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/ir_simplifier.cpp:3117]: Fixed.

**Pull request closed without merging**